File: c++/temp_net_parser.cpp
#include <iostream>
#include <vector>
#include <string>
#include <utility>
#include <fstream>
#include <map>

//Collection of functions for parseing time stamped edge list and converting to format for memory estimation.


std::vector<int> parse_edge_string(std::string string_edge, char seperator) {
	
	int string_index=0;
	int seperator_index=0;
	std::vector<std::string> split_string_edge(3);	//Only take three strings.
	std::vector<int> edge_vector(3);
	
	//Loop over each element of the string_edge.
	while(string_index < string_edge.length()) {
		//If the element is the seperator character, then move to the next element in the split vector.
		//Do not include seperator in the stored strings.
		//Else add the element to the current string in split vector.
		if(string_edge[string_index] == seperator) {
			seperator_index++;
		}
		else {
			split_string_edge[seperator_index] += string_edge[string_index];
		}
		string_index++;
	}
	
	//Convert strings for each edge/time stamp to integers for output vector.
	for(seperator_index=0 ; seperator_index<3 ; seperator_index++) {
		edge_vector[seperator_index] = stoi(split_string_edge[seperator_index]);
	}
	
	return edge_vector;
}


//Convert a time stamped edge list to a vector of time series for each edge.
//Each time series represents a single edge.
//Each element in a time series is a bool, 1 for link present, 0 else. 
std::vector<std::vector<bool> > convert_edge_list_format(std::vector<std::vector<int> > int_edge_list) {
	
	int edge_index;
	int unique_counter = 0;							//Count the number of unique edges.
	int end_time = int_edge_list.back()[2];		//Get the final time stamp for the edge list.
	std::pair<int,int> edge;
	
	std::vector<bool> dum_edge_series(end_time,0);
	
	std::vector<std::vector<bool> > bool_edge_list;			//Output format.
	std::map<std::pair<int, int>, int> edge_ind_map;		//Label each unique edge.
	
	for(edge_index=0 ; edge_index<int_edge_list.size() ; edge_index++) {
		edge.first = int_edge_list[edge_index][0];
		edge.second = int_edge_list[edge_index][1]; 
		
		//See if we can find the current edge in the list of previously seen edges.
		std::map<std::pair<int,int>, int>::const_iterator e_map_iter = edge_ind_map.find(edge);
		if (e_map_iter != edge_ind_map.end()) {
			//If we have seen this edge before, update the output.
			bool_edge_list[e_map_iter->second][int_edge_list[edge_index][2]] = 1;
		}
		else {
			//If this edge is new to us, then make a new element in the output.
			edge_ind_map[edge] = unique_counter;
			bool_edge_list.push_back(dum_edge_series);
			bool_edge_list[unique_counter][int_edge_list[edge_index][2]] = 1;
			unique_counter++;
		}
	}
	return bool_edge_list;
}


//Function to get time stamped edge list from a file.
//Requires the path to the file containing the edge list data and the seperator for each column in the file.
//Optional parameter: flag for header line, if true first line of file will be ignored.
std::vector<std::vector<bool> > parse_edge_file(std::string file_path, char seperator, bool has_headline){
	
	std::ifstream inFile;
	inFile.open(file_path);
	
	if (!inFile) {
		std::cerr << "Unable to open file.";
		exit(1);   // call system to stop
	}
	
	//If the function is told that the file will have a header line, then ignore it.
	if(has_headline) {
		inFile.ignore(1,'\n');
	}
	
	std::vector<std::vector<int> > parsed_edge_list;
	std::vector<int> dum_edge(3);						//Vector to hold each time stamped edge that appears.
	std::string line = "";
	while(getline(inFile,line)) {
		dum_edge = parse_edge_string(line, seperator);		//Parse the string for this edge as a vector of ints.
		parsed_edge_list.push_back(dum_edge);		//Add the parsed edge to the edge list.
	}
	
	std::cout << "Got edge list, now formatting..." << std::endl;
		
	//Convert the parsed edge list into the vec<vec<bool>> format used for finding memory.
	std::vector<std::vector<bool> > formatted_edge_list = convert_edge_list_format(parsed_edge_list);
	
	return formatted_edge_list;
}

File: python_2/from_file/temp_net_parser.py
#!/usr/bin/python
import numpy as np

def main():
	
	print("Program build correctly!")
	print("Note, this program is not intended to run on its own...")
	print("This is the only output.")
	
	return


"""
Function to get time stamped edge list from a file.
Requires the path to the file containing the edge list data and the seperator for each column in the file.
Optional parameter: flag for header line, if true first line of file will be ignored.
"""
def parse_edge_file(file_path, seperator, has_header=False):
	
	f = open(file_path,'r')
	
	if has_header==True:
		f.readline()		#If there is a header line, then skip it.
	
	
	edge_list = []
	
	for line in f:
		edge_list.append(parse_edge_string(line, seperator))	#Parse the edge and put it in a list.
	
	formatted_edge_list = convert_edge_list_format(edge_list)
	
	return formatted_edge_list
	
	
"""
Take the string representing an edge in the edge list file and return an appropriate format.
Needs the seperator character used (given by parse_edge_file function).
"""
def parse_edge_string(edge_string, seperator):
	
	#To ensure that any seperator can be used, work with the raw representation of the string.
	#This is a bit of a hack.
	split_string = repr(edge_string)[1:-1].split(seperator)
		
	#Get the nodes and the time stamp, ignoring the end of line character.
	source_node = int(split_string[0])
	target_node = int(split_string[1])
	try:
		time_stamp = int(split_string[2][0:-2])	
	except ValueError:
		time_stamp = int(split_string[2])	
	
	return [source_node,target_node,time_stamp]
	
	
"""
Convert a time stamped edge list to a vector of time series for each edge.
//Each time series represents a single edge.
This method is in no way optimal, but should hopefully be clear.
"""
def convert_edge_list_format(edge_list):
	
	end_time = edge_list[-1][2]		#Get the last time stamp value in the edge list.
	
	#First get the number of unique edges.
	#NOTE: we do not assume the network is undirected.
	seen_edges = {}			#Dict of all unique edges and their numeric label.
	edge_count = 0
	for t_edge in edge_list:
		edge = (t_edge[0], t_edge[1])			#Just take the node labels.
		if edge not in seen_edges.keys():
			seen_edges[edge] = edge_count
			edge_count +=1					
			
	num_edges = edge_count		#Get number of edges.
	
	#Initialise array for holding time series.
	#edge_series = np.zeros((num_edges, end_time+1), dtype=bool)
	edge_series = [[0]*(end_time+1)]*num_edges
	
	#Fill array.
	for t_edge in edge_list:
		edge = (t_edge[0], t_edge[1])
		edge_series[seen_edges[edge]][t_edge[2]] = 1			#Update the edge series array.
	
	return edge_series



if __name__ == "__main__":
	main()